### PR TITLE
Security 1.0

### PR DESCRIPTION
**Pull request by aksoli666**
added User entity and required dto`s
added method register and have repo, service, controller layers used data validation
added @FieldMatcher
added RegistrationException

### CODE REVIEW

File: src/main/java/shop/annatation/FieldMatch.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package shop.annatation;
```
Comment (Elena-Bruyako) [inline src/main/java/shop/annatation/FieldMatch.java]:
```suggestion
package shop.validation;
```

File: src/main/java/shop/validator/FieldMatchValidator.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
+package shop.validator;
```
Comment (Elena-Bruyako) [inline src/main/java/shop/validator/FieldMatchValidator.java]:
```suggestion
package shop.validation;
```

File: src/main/java/shop/validator/FieldMatchValidator.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    }
+
+    @Override
+    public boolean isValid(String field, ConstraintValidatorContext constraintValidatorContext) {
+        String firstFieldValue = (String) new BeanWrapperImpl(field)
+                .getPropertyValue(firstFieldName);
+        String secondFieldValue = (String) new BeanWrapperImpl(field)
+                .getPropertyValue(secondFieldName);
+
+        if (firstFieldValue != null && secondFieldValue != null) {
```
Comment (Elena-Bruyako) [inline src/main/java/shop/validator/FieldMatchValidator.java]: You can use Objects.equals here

File: src/main/java/shop/dto/request/RegisterUserRequestDto.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+import shop.annatation.FieldMatch;
+
+@Setter
+@Getter
+@FieldMatch(
+        firstFieldName = "password",
+        secondFieldName = "repeatPassword",
+        message = "Password must match!"
+)
+public class RegisterUserRequestDto {
```
Comment (Elena-Bruyako) [inline src/main/java/shop/dto/request/RegisterUserRequestDto.java]:
```suggestion
public class UserRegistrationRequestDto {
```

File: src/main/java/shop/dto/responce/RegisterUserResponseDto.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
+package shop.dto.responce;
+
+import jakarta.validation.constraints.Email;
+import jakarta.validation.constraints.NotBlank;
+import lombok.Getter;
+import lombok.Setter;
+
+@Setter
+@Getter
+public class RegisterUserResponseDto {
```
Comment (Elena-Bruyako) [inline src/main/java/shop/dto/responce/RegisterUserResponseDto.java]:
```suggestion
public class UserResponseDto {
```

File: src/main/java/shop/dto/responce/RegisterUserResponseDto.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    @NotBlank
```
Comment (Elena-Bruyako) [inline src/main/java/shop/dto/responce/RegisterUserResponseDto.java]: don't need validate response

File: src/main/java/shop/entity/User.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+            unique = true,
+            nullable = false
+    )
+    private String email;
+    @Column(nullable = false)
+    private String password;
+    @Column(
+            name = "first_name",
+            nullable = false
+    )
```
Comment (Elena-Bruyako) [inline src/main/java/shop/entity/User.java]:
```suggestion
    @Column(nullable = false)
```

File: src/main/java/shop/entity/User.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    private String firstName;
+    @Column(
+            name = "last_name",
+            nullable = false
+    )
```
Comment (Elena-Bruyako) [inline src/main/java/shop/entity/User.java]:
```suggestion
    @Column(nullable = false)
```

File: src/main/java/shop/entity/User.java [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    private String lastName;
+    @Column(name = "shipping_address")
```
Comment (Elena-Bruyako) [inline src/main/java/shop/entity/User.java]:
```suggestion
```

File: src/main/java/shop/entity/User.java [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+@Getter
+@Setter
+public class User {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(
+            unique = true,
+            nullable = false
+    )
```
Comment (Elena-Bruyako) [inline src/main/java/shop/entity/User.java]:
```suggestion
    @Column(unique = true, nullable = false)
```

File: src/main/java/shop/service/impl/UserServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+@Service
+@RequiredArgsConstructor
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final UserMapper userMapper;
+
+    @Override
+    public RegisterUserResponseDto register(RegisterUserRequestDto registerUserRequestDto)
+            throws RegistrationException {
+        if (userRepository.findByEmail(registerUserRequestDto.getEmail()).isPresent()) {
```
Comment (Elena-Bruyako) [inline src/main/java/shop/service/impl/UserServiceImpl.java]:
```suggestion
        if (userRepository.existsByEmail(registerUserRequestDto.getEmail())) {
```

File: src/main/java/shop/service/impl/UserServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+            throw new RegistrationException("Can't register user with given email");
+        }
+        User user = userMapper.toUser(registerUserRequestDto);
+        User savedUser = userRepository.save(user);
+        return userMapper.toRegisterUserResponseDto(savedUser);
```
Comment (Elena-Bruyako) [inline src/main/java/shop/service/impl/UserServiceImpl.java]:
```suggestion
        userRepository.save(user);
        return userMapper.toRegisterUserResponseDto(user);
```

File: src/main/resources/db/changelog/changes/02.-create-users-table.yaml [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+                  constraints:
+                    nullable: false
+              - column:
+                  name: last_name
+                  type: varchar(255)
+                  constraints:
+                    nullable: false
+              - column:
+                  name: shipping_address
+                  type: varchar(255)
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/02.-create-users-table.yaml]: did you add 02 changelog to master changelog? 

Add empty line at the end of file

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by ahoienko — APPROVED** [review]
Good job!

**Pull request merged**